### ISSUE TITLE
Windows Server 2019 (ECS optimized AMI) - ECS task role doesn't work

### ISSUE THREAD

**Issue by brubtsov** [aws/amazon-ecs-agent#2105]
### Summary
I'm running an instance based on Win Server 2019 ECS optimized AMI. It seems that attached ECS task role does not work. 

### Description
My container fails with error
`Error when retrieving credentials from container-role: Error retrieving metadata: Received error when attempting to retrieve ECS metadata: Connect timeout on endpoint URL: "http://169.254.170.2/v2/credentials/a3b23aef-3271-4d34-9a78-b6c4f2fe6de8"`
during download from S3.
I use AMI Windows_Server-2019-English-Full-ECS_Optimized-2019.05.10 . This issue is similar to #1644 that I created some time ago. The difference is that now I'm using official AMI for Win Server 2019 instead of custom AMI.

- Base Docker image - mcr.microsoft.com/dotnet/framework/runtime:4.7.2-windowsservercore-ltsc2019

- At the beginning, when container starts we run script https://docs.aws.amazon.com/AmazonECS/latest/developerguide/windows_task_IAM_roles.html to initialize IAM role feature

- We use AWS CLI to download from S3 - https://s3.amazonaws.com/aws-cli/AWSCLI64PY3.msi

### Expected Behavior
ECS task work correctly: container is able to use role and download file from S3.

### Observed Behavior
See 'Description'.

### Environment Details
t3.medium instance created from AMI Windows_Server-2019-English-Full-ECS_Optimized-2019.05.10 (ami-05da69b2d804943e6).
User data:
```
<powershell>
[Environment]::SetEnvironmentVariable("ECS_ENABLE_CPU_UNBOUNDED_WINDOWS_WORKAROUND", "true", "Machine")
[Environment]::SetEnvironmentVariable("ECS_DISABLE_METRICS", "false", "Machine")
Import-Module ECSTools
Initialize-ECSAgent -Cluster 'Win2k19' -EnableTaskIAMRole
</powershell>
```

Basically for Win Server 2019 cluster I use the same setup (roles, VPC configuration, Dockerfile, etc. ) as I use for Win Server 2016 cluster (where everything works as expected).

### Supporting Log Snippets
I can provide ECS logs if needed. ECS Agent setup log contains no errors or warnings. ECS Agent operational logs have only few minor warnings that are probably not related to this issue.

**Comment by brubtsov** [comment]
I managed to get it work by creating another cluster via 'Create Cluster' wizard. New instance has been created in the same subnet from the same AMI and it's running same task definition.
The only difference is that "good" instance (where task role works) is m5.large and "bad" instance is t3.medium.

**Comment by sharanyad** [comment]
@brubtsov how frequently do you face this issue? how many tasks approximately are running at the time this issue came up? If you can email us the debug level agent logs, it would be easier to debug this particular issue. You can send them over at sharanyd at amazon.com .
Thanks!

**Comment by brubtsov** [comment]
@sharanyad when I first time encountered this problem, it happened for every single task. At that time there was only one task in a cluster with single t3.medium instance.
After I launched new set of instances in different cluster - problem disappeared. But not completely - I still see this 'Connect timeout on endpoint URL' error once per 30 - 40 tasks. I changed log level and I'll try to collect some more logs.

Also today I saw another problem - ECS Agent doesn't clean up images and as a result Docker consumed all disk space. I don't know, should I create separate issue for this one or not.

**Comment by sharanyad** [comment]
@brubtsov Could you try increasing the value for `ECS_TASK_METADATA_RPS_LIMIT` for steady state and burst throttle limits for the task metadata endpoint? Current default value is `40,60` which may need to be increased for your use case. More information on it [here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-agent-config.html). 

Wrt agent not cleaning up images, could you please make sure the image cleanup is not turned off, and `ECS_IMAGE_CLEANUP_INTERVAL` , `ECS_IMAGE_MINIMUM_CLEANUP_AGE` is not set to a higher value?

**Comment by adnxn** [comment]
@brubtsov: are you unblocked on this issue? were you able to apply @sharanyad's suggestion?

**Comment by brubtsov** [comment]
@adnxn I tried to play with `ECS_TASK_METADATA_RPS_LIMIT` parameter but it didn't help. Currently we use workaround with passing AWS credentials via task definition environment variables.

Some time ago I sent ECS Agent logs to email provided above, but I didn't get any answer since then.

Also, it look like ECS Task Metadata Endpoint doesn't work at all inside our Windows containers. So it's not just a credentials / task role problem.

**Comment by adnxn** [comment]
@brubtsov: do you mean you're not able to rely on consistent behavior using task iam roles? im looking at your earlier message, below

> After I launched new set of instances in different cluster - problem disappeared. But not completely - I still see this 'Connect timeout on endpoint URL' error once per 30 - 40 tasks. I changed log level and I'll try to collect some more logs.

> Connect timeout on endpoint URL

we've seen this type of intermittent failure resulting from constrained resources on instances, and likely path forward would be to either size up the instances backing your cluster (which seems like you've tried with m5.large) or increase the number of instances backing your cluster and reduce the load put on each instance by spreading tasks out.

**Comment by adnxn** [comment]
closing this issue since we havent heard back, please feel free to reopen if you need and update with new information as you see fit.

**Comment by smokentar** [comment]
For anyone that encounters this in the future:

Situation is caused by windows dynamic port reservation.

AWS ECS tasks using task IAM roles for authorizing APIs will query the following endpoint to retrieve credentials:
http://169.254.170.2/v2/credentials

This will be done on port 51679.

In **ProgramData\Amazon\ECS\logs\** we can observe:
```
Error running introspection endpoint: listen tcp :51678: bind: An attempt was made to access a socket in a way forbidden by its access
permissions.

Error running task api: listen tcp 127.0.0.1:51679: bind: An attempt was made to access a socket in a way forbidden by its access permi
ssions.
```

On the faulty server hosting docker containers run `netsh interface ipv4 show excludedportrange protocol=tcp`
This will display a list of ports that have been randomly reserved and cannot be bound to.
The list would show a range similar to 51644 : 51743 (always random)

This is intermittent as sometimes another range would be excluded and those ports will be unaffected. 
Reproducing is entirely based on chance.

To reset and exclude required ports without restart:
```
net stop winnat
stop-service hns

netsh int ipv4 add excludedportrange protocol=tcp startport=51678 numberofports=1
netsh int ipv4 add excludedportrange protocol=tcp startport=51679 numberofports=1

net start winnat
start-service hns
```

Run `netsh interface ipv4 show excludedportrange protocol=tcp` and you will see different ranges + the required ports excluded from dynamic reservation.
Settings will persist through a reboot.
